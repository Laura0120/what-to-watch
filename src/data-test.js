export const movie = {
  id: 1,
  poster: `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/Macbeth.jpg`,
  preview: `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/macbeth.jpg`,
  backgroundImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/Macbeth.jpg`,
  backgroundColor: `#F1E9CE`,
  video: `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`,
  previewVideo: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
  title: `Macbeth`,
  rating: {
    ratingScore: 3.3,
    countVotesRating: 48798
  },
  director: `Justin Kurzel`,
  starring: [`Michael Fassbender`, `Marion Cotillard`, `Jack Madigan`],
  year: 2015,
  runtime: 113,
  genre: `Drama`,
  description: `Macbeth, the Thane of Glamis, receives a prophecy from a trio of witches that one day he will become King of Scotland. Consumed by ambition and spurred to action by his wife, Macbeth murders his king and takes the throne for himself.`,
  isFavorite: false
};


export const noop = () => {};

export const comment = {
  comment: `A movie that will take you to another world full of emotions.`,
  date: `2020-11-03T13:38:44.769Z`,
  id: 1,
  rating: 8.6
};
