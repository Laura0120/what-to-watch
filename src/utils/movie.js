const getRatingLevel = (ratingNumber) => {
  if (ratingNumber >= 0.0 && ratingNumber < 3.0) {
    return `Bad`;
  } else if (ratingNumber >= 3.0 && ratingNumber < 5.0) {
    return `Normal`;
  } else if (ratingNumber >= 5.0 && ratingNumber < 8.0) {
    return `Good`;
  } else if (ratingNumber >= 8.0 && ratingNumber < 10.0) {
    return `Very good`;
  } else {
    return `Awesome`;
  }
};

export {getRatingLevel};
