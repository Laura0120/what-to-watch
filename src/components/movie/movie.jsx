import React, {useEffect} from 'react';
import {connect} from "react-redux";
import PropTypes from 'prop-types';
import {matchPath} from "react-router-dom";

import {FUNCTION, STRING, MOVIES, MOVIE, COMMENTS} from '../../prop-type';
import {AuthorizationStatus, AppRoute} from "../../const";
import {ActionCreator} from "../../store/action";
import {changeFavorite, fetchFavorite, fetchMovieById, fetchCommentsByMovieId} from "../../store/api-actions";
import MoreLikeThis from '../more-like-this/more-like-this';
import MovieButtons from '../movie-buttons/movie-buttons';
import UserBlock from '../user-block/user-block';
import {getMoviesSimilar} from '../../store/selectors';
import {adaptToClientMovie} from '../../utils/adapt';

const getAddReviewLink = (status, callback, id) => {
  return (
    status === AuthorizationStatus.AUTH ?
      <a href='' className='btn movie-card__button' onClick={(evt)=>callback(evt, id)} >
        Add review
      </a> : ``);
};

const Movie = (props) => {
  const {movie, loadMovie, location, renderTabs, comments, moviesLikeThis, authorizationStatus,
    onAddReviewClick, onPlayClick, onFavoriteClick, onMainPageClick, onMyListButtonClick, onMovieClick} = props;
  const {poster, backgroundImage, title, year, genre, id, isFavorite} = movie;

  useEffect(() => {
    const locationInfo = matchPath(location.pathname, {
      path: AppRoute.FILM_ID,
      exact: true,
      strict: true
    });

    if (movie.id !== Number(locationInfo.params.id)) {
      loadMovie(locationInfo.params.id);
    }
  }, []);

  return (
    <React.Fragment>
      <section className='movie-card movie-card--full'>
        <div className='movie-card__hero'>
          <div className='movie-card__bg'>
            <img src={backgroundImage} alt={title} />
          </div>

          <h1 className='visually-hidden'>WTW</h1>

          <header className='page-header movie-card__head'>
            <div className='logo'>
              <a href='#'
                className='logo__link'
                onClick={(evt) => {
                  evt.preventDefault();
                  onMainPageClick();
                }}>
                <span className='logo__letter logo__letter--1'>W</span>
                <span className='logo__letter logo__letter--2'>T</span>
                <span className='logo__letter logo__letter--3'>W</span>
              </a>
            </div>

            <UserBlock authorizationStatus={authorizationStatus} onMyListButtonClick={onMyListButtonClick}/>
          </header>
          <div className='movie-card__wrap'>
            <div className='movie-card__desc'>
              <h2 className='movie-card__title'>{title}</h2>
              <p className='movie-card__meta'>
                <span className='movie-card__genre'>{genre}</span>
                <span className='movie-card__year'>{year}</span>
              </p>

              <MovieButtons id={id} isFavorite={isFavorite} onPlayClick={onPlayClick} onFavoriteClick={onFavoriteClick}
                authorizationStatus={authorizationStatus} onAddReviewClick={onAddReviewClick} getAddReviewLink={getAddReviewLink}/>
            </div>
          </div>
        </div>

        <div className='movie-card__wrap movie-card__translate-top'>
          <div className='movie-card__info'>
            <div className='movie-card__poster movie-card__poster--big'>
              <img src={poster} alt={title} width='218' height='327' />
            </div>

            {renderTabs(movie, comments)}

          </div>
        </div>
      </section>

      <div className='page-content'>
        <MoreLikeThis movies={moviesLikeThis} onMovieClick={onMovieClick}/>
        <footer className='page-footer'>
          <div className='logo'>
            <a href='#'
              className='logo__link logo__link--light'
              onClick={(evt) => {
                evt.preventDefault();
                onMainPageClick();
              }}>
              <span className='logo__letter logo__letter--1'>W</span>
              <span className='logo__letter logo__letter--2'>T</span>
              <span className='logo__letter logo__letter--3'>W</span>
            </a>
          </div>

          <div className='copyright'>
            <p>Â© 2019 What to watch Ltd.</p>
          </div>
        </footer>
      </div>
    </React.Fragment>
  );
};

Movie.propTypes = {
  movie: MOVIE.isRequired,
  location: PropTypes.object.isRequired,
  moviesLikeThis: MOVIES.isRequired,
  renderTabs: FUNCTION,
  comments: COMMENTS.isRequired,
  authorizationStatus: STRING,
  onAddReviewClick: FUNCTION,
  onPlayClick: FUNCTION,
  onFavoriteClick: FUNCTION,
  onMyListButtonClick: FUNCTION,
  onMovieClick: FUNCTION,
  onMainPageClick: FUNCTION,
  loadMovie: FUNCTION,
};

const mapStateToProps = (state) => ({
  moviesLikeThis: getMoviesSimilar(state),
  movie: adaptToClientMovie(state.DATA.openedMovie),
  comments: state.DATA.openMovieComments,
  authorizationStatus: state.USER.authorizationStatus,
});

const mapDispatchToProps = (dispatch) => ({
  onAddReviewClick(evt, id) {
    evt.preventDefault();
    dispatch(ActionCreator.redirectToRoute(`${id}/review`));
  },
  onPlayClick(id) {
    dispatch(ActionCreator.redirectToRoute(`/player/${id}`));
  },
  onFavoriteClick(id, isFavorite) {
    dispatch(changeFavorite(id, isFavorite));
  },
  onMyListButtonClick() {
    dispatch(fetchFavorite());
    dispatch(ActionCreator.redirectToRoute(AppRoute.MY_LIST));
  },
  loadMovie(id) {
    dispatch(fetchMovieById(id));
    dispatch(fetchCommentsByMovieId(id));
  },
  onMovieClick(id) {
    dispatch(ActionCreator.redirectToRoute(`/films/${id}`));
    dispatch(fetchMovieById(id));
    dispatch(fetchCommentsByMovieId(id));
  },
  onMainPageClick() {
    dispatch(ActionCreator.redirectToRoute(`/`));
  }
});
export {Movie};
export default connect(mapStateToProps, mapDispatchToProps)(Movie);
