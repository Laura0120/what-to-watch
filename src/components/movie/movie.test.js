import React from 'react';
import renderer from 'react-test-renderer';

import {Movie} from './movie';
import {movie, noop} from '../../data-test';

it(`Movie is rendered correctly`, () => {
  const tree = renderer.create(
      <Movie
        movie={movie}
        comments={[]}
        moviesLikeThis={[movie]}
        authorizationStatus = {`AUTH`}
        onAddReviewClick={noop}
        onPlayClick={noop}
        onFavoriteClick={noop}
        renderTabs={noop}
        onMyListButtonClick={noop}
        onMovieClick={noop}
        loadMovie={noop}
        onMainPageClick = {noop}
        location={{pathname: `/films/1`}}
      />
  ).toJSON();

  expect(tree).toMatchSnapshot();
});
