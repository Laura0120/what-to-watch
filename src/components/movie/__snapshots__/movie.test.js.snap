// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movie is rendered correctly 1`] = `
Array [
  <section
    className="movie-card movie-card--full"
  >
    <div
      className="movie-card__hero"
    >
      <div
        className="movie-card__bg"
      >
        <img
          alt="Macbeth"
          src="https://assets.htmlacademy.ru/intensives/javascript-3/film/background/Macbeth.jpg"
        />
      </div>
      <h1
        className="visually-hidden"
      >
        WTW
      </h1>
      <header
        className="page-header movie-card__head"
      >
        <div
          className="logo"
        >
          <a
            className="logo__link"
            href="#"
            onClick={[Function]}
          >
            <span
              className="logo__letter logo__letter--1"
            >
              W
            </span>
            <span
              className="logo__letter logo__letter--2"
            >
              T
            </span>
            <span
              className="logo__letter logo__letter--3"
            >
              W
            </span>
          </a>
        </div>
        <div
          className="user-block"
        >
          <div
            className="user-block__avatar"
            onClick={[Function]}
          >
            <img
              alt="User avatar"
              height="63"
              src="img/avatar.jpg"
              width="63"
            />
          </div>
        </div>
      </header>
      <div
        className="movie-card__wrap"
      >
        <div
          className="movie-card__desc"
        >
          <h2
            className="movie-card__title"
          >
            Macbeth
          </h2>
          <p
            className="movie-card__meta"
          >
            <span
              className="movie-card__genre"
            >
              Drama
            </span>
            <span
              className="movie-card__year"
            >
              2015
            </span>
          </p>
          <div
            className="movie-card__buttons"
          >
            <button
              className="btn btn--play movie-card__button"
              onClick={[Function]}
              type="button"
            >
              <svg
                height="19"
                viewBox="0 0 19 19"
                width="19"
              >
                <use
                  xlinkHref="#play-s"
                />
              </svg>
              <span>
                Play
              </span>
            </button>
            <button
              className="btn btn--list movie-card__button"
              onClick={[Function]}
              type="button"
            >
              <svg
                height="20"
                viewBox="0 0 19 20"
                width="19"
              >
                <use
                  xlinkHref="#add"
                />
              </svg>
              <span>
                My list
              </span>
            </button>
            <a
              className="btn movie-card__button"
              href=""
              onClick={[Function]}
            >
              Add review
            </a>
          </div>
        </div>
      </div>
    </div>
    <div
      className="movie-card__wrap movie-card__translate-top"
    >
      <div
        className="movie-card__info"
      >
        <div
          className="movie-card__poster movie-card__poster--big"
        >
          <img
            alt="Macbeth"
            height="327"
            src="https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/Macbeth.jpg"
            width="218"
          />
        </div>
      </div>
    </div>
  </section>,
  <div
    className="page-content"
  >
    <section
      className="catalog catalog--like-this"
    >
      <h2
        className="catalog__title"
      >
        More like this
      </h2>
      <div
        className="catalog__movies-list"
      >
        <article
          className="small-movie-card catalog__movies-card"
          id={1}
          onClick={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
        >
          <div
            className="small-movie-card__image"
          >
            <video
              height="175"
              muted={true}
              poster="https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/macbeth.jpg"
              src="https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4"
              width="280"
            />
          </div>
          <h3
            className="small-movie-card__title"
          >
            <a
              className="small-movie-card__link"
              href="#"
            >
              Macbeth
            </a>
          </h3>
        </article>
      </div>
      
    </section>
    <footer
      className="page-footer"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link logo__link--light"
          href="#"
          onClick={[Function]}
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        className="copyright"
      >
        <p>
          Â© 2019 What to watch Ltd.
        </p>
      </div>
    </footer>
  </div>,
]
`;
